lib/common/src/Url.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+namespace AmpProject;
+
+/**
+ * Helper class to work with URLs.
+ *
+ * @package ampproject/common
+ */
+final class Url
+{
+
+    /**
+     * Default URL parts to use when constructing an absolute URL out of a relative one.
+     *
+     * @var string[]
+     */
+    const URL_DEFAULT_PARTS = [
+        'scheme'   => 'https',
+        'host'     => 'example.com',
+        'port'     => '',
+        'user'     => '',
+        'pass'     => '',
+        'path'     => '',
+        'query'    => '',
+        'fragment' => '',
+    ];
+
+    /**
+     * Check whether a given src string is a valid image source URL.
+     *
+     * @param string $src Src string to validate.
+     * @return bool Whether the src string is a valid image source URL.
+     */
+    public static function isValidImageSrc($src)
+    {
+        list($scheme, $host, $port, $user, $pass, $path, $query, $fragment) = array_values(
+            array_merge(
+                self::URL_DEFAULT_PARTS,
+                (array)parse_url($src)
+            )
+        );
+
+        if ($scheme === 'data') {
+            return false;
+        }
+
+        $userpass = $user;
+
+        if (! empty($pass)) {
+            $userpass = "{$user}:{$pass}";
+        }
+
+        if (! empty($userpass)) {
+            $userpass .= '@';
+        }
+
+        $url = sprintf(
+            '%s://%s%s%s/%s%s%s',
+            $scheme,
+            $host,
+            $userpass,
```

Comment:
Shouldn't it be `$userpass` then `$host`?

```suggestion
            $userpass,
            $host,
```